File: ws-gateway/internal/observability/logger.go
package observability

import (
	"context"
	"os"

	"go.opentelemetry.io/otel/trace"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

var Logger *zap.Logger

func initLogger() {
	config := zap.NewProductionConfig()
	config.EncoderConfig.TimeKey = "timestamp"
	config.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder

	serviceName := os.Getenv("OTEL_SERVICE_NAME")
	if serviceName == "" {
		serviceName = "ws-gateway"
	}

	var err error
	Logger, err = config.Build(zap.Fields(
		zap.String("service_name", serviceName),
	))
	if err != nil {
		panic(err)
	}
}

func GetLoggerWithTrace(ctx context.Context) *zap.Logger {
	span := trace.SpanFromContext(ctx)
	if span.SpanContext().IsValid() {
		return Logger.With(
			zap.String("trace_id", span.SpanContext().TraceID().String()),
			zap.String("span_id", span.SpanContext().SpanID().String()),
		)
	}
	return Logger
}

File: api-gateway/src/observability/logger.ts
import pino from 'pino';
import { trace, context } from '@opentelemetry/api';

export const logger = pino({
  level: process.env.LOG_LEVEL || 'info',
  mixin() {
    const span = trace.getSpan(context.active());
    if (span) {
      const spanContext = span.spanContext();
      return {
        trace_id: spanContext.traceId,
        span_id: spanContext.spanId,
        service_name: process.env.OTEL_SERVICE_NAME || 'api-gateway',
      };
    }
    return {
      service_name: process.env.OTEL_SERVICE_NAME || 'api-gateway',
    };
  },
});

// Replaces console.log with structured logging
export const instrumentConsole = () => {
  console.log = (...args) => logger.info(args.length > 1 ? args : args[0]);
  console.info = (...args) => logger.info(args.length > 1 ? args : args[0]);
  console.error = (...args) => logger.error(args.length > 1 ? args : args[0]);
  console.warn = (...args) => logger.warn(args.length > 1 ? args : args[0]);
};
